### PR TITLE
Prototype 1: New "Enlarge Canvas" UI

### PR DESCRIPTION
**Pull request by Minecraftian14**
Done:
* Enlarge Canvas Menu is working.
* The new UI seems to be working just fine...
* It is showing a "very" rough representation of the existing canvas and new canvas details.

Todo:
I would like it to show the current image portions and the hidden image revealing portions instead of some red/blue boxes.

Note:
So basically, I want to know how to get an instance of the current image or the whole image with it's hidden parts, so that I can draw it in the preview box.

The UI is not finalized, all improvements accepted 😅.
There is that issue with a "red" line showing even when all sliders are 0, dunno why, but maybe because of float to int conversion.

Screenshots:

### CODE REVIEW

File: src/main/java/pixelitor/gui/utils/SliderSpinner.java [hunk 1]
```diff
@@ -59,17 +63,28 @@
     private boolean spinnerMoved = false;
 
     public SliderSpinner(RangeParam model, TextPosition position, boolean addDefaultButton) {
-        this(model, null, null, position, addDefaultButton);
+        this(model, position, addDefaultButton, HORIZONTAL);
```
Comment (Minecraftian14) [inline src/main/java/pixelitor/gui/utils/SliderSpinner.java:66]: Yeah, I did a mess with the constructors 😅.
:wasn't sure how should I resolve that...

**Comment by Minecraftian14** [comment]
Fixed the red line issue...
image

**Comment by lbalazscs** [comment]
Wow, this looks even better than what I imagined!! (I didn't think of vertical sliderspinners) (I didn't check the code yet, but I'm on it)

**Comment by Minecraftian14** [comment]
😁
Just tell me what I can improve, and I am on it 👍.

**Comment by Minecraftian14** [comment]
What i actually wanted, i dont know if it's possible...
https://user-images.githubusercontent.com/52451860/119494843-5fc25680-bd7f-11eb-8ead-bb9b20ed250a.png

**Comment by lbalazscs** [comment]
Overall I'm very impressed with your code, and if you say you are done, I am happy to merge into the master branch! Here are some minor observations, but these aren't obstacles of accepting the pull request:

1) The ResizeListener idea is also very interesting, but the problem is that as you probably noticed, currently the sliderspinners don't behave well, when there isn't enough space. 

2) I don't understand the line

```java
                    // Gives the part of image placed within the canvas
                    OpenImages.getActiveComp().getCompositeImage();
```
in ResizeListener, because the return value is ignored.

3) If you are reformatting the part of code you don't modify, then it makes it harder to me to look at the diffs, because there is formatting noise. (I don't care about your code style, only about clean diffs) What IDE are you using? I use Intellij Idea (the free community version is fine for Pixelitor), and it has the feature of not formatting automatically unmodified code. Anyway, if you just turn off auto-formatting, it should be OK.

4) The red and blue colors are unusual in the Pixelitor GUI, I try to use a small number of colors. No big deal, this is easy to change at any time.

**Comment by Minecraftian14** [comment]
"The ResizeListener idea is also very interesting, but the problem is that as you probably noticed, currently the sliderspinners don't behave well, when there isn't enough space."

Oh! I didn't notice it! How do I reproduce it?
ResizeListener is there just in case to prevent drawing the CanvasEditor if it's initialised but not needed.
I think I can actually remove it!

"I don't understand the line"
Haha, my fault, i was wondering how to get the image!
And forgot to remove that - It's removed now.

"If you are reformatting the part of code you don't modify, then it makes it harder to me to look at the diffs,"
Yeah, I also found it difficult when I went to analyze the changes myself.
Will keep that in mind next time 😅

"The red and blue colors are unusual in the Pixelitor GUI, I try to use a small number of colors. No big deal, this is easy to change at any time."
Yeah, in reality they were just for testing purpose...

**Comment by lbalazscs** [comment]
I'm not sure I understand the "What i actually wanted" image... Do you want to show the full layer content size? contentLayer.getContentBounds() can give you that 

Or the full layer contents? (currently there are no utility methods for that)

**Comment by Minecraftian14** [comment]
Ah sorry, yeah i just updated the comment with an image

**Comment by Minecraftian14** [comment]
Red box: Current Canvas
Blue box: New Canvas....

Yeah! Colors can be changed

image

**Comment by Minecraftian14** [comment]
So is this possible?
If yes please gimme a few hints 😅

**Comment by lbalazscs** [comment]
That's a nice idea, but it doesn't seem very important, because I think most of the time when people enlarge the canvas, there will be nothing there, just empty pixels. Also there's no easy way to do it. So my suggestion is to leave it as it is for now, and perhaps you can change it in another pull request, if you want.

**Comment by lbalazscs** [comment]
You would have to calculate the "full" composite image, but this is pretty complicated (different kinds of layers, masks, translations), and the current code everywhere assumes that the composite image has the size of the canvas. A hack would be to temporarily change the canvas size to the "full" value, but this seems dangerous.

On the other hand, just showing the "full" size of the canvas (perhaps with a third color) should be easier: all content layers could be iterated for their getContentBounds(), and using Rectangle.union on the results gives the  canvas size that corresponds to all the content. 

But as I said, even this feature would be only rarely useful, because usually there is no content outside the current canvas. The pull request is already great as it is, let me know if it should be merged or if you still plan to change something.

**Comment by Minecraftian14** [comment]
Thanks a lot for giving me this opportunity 😀...

Surprisingly I completed it in just half day? while I was asking for a week 🤣

**Comment by lbalazscs** [comment]
Well, if you are interested in further contributing to Pixelitor, there are plenty of other opportunities... (I'm thinking about concrete ideas, but right now I have to go. Perhaps you also noticed something that would be fun to improve?)
 
I missed earlier your reply to ResizeListener. The issue is that if I resize the whole "Enlarge Canvas" dialog, the preview gets resized, which is cool, but the sliderspinners don't like when the dialog gets smaller, and not bigger. I thought that you did the ResizeListener because you wanted to support window resizing, just like the Navigator window supports resizing. It's no big deal, probably most users won't even get the idea of resizing the window.

**Comment by Minecraftian14** [comment]
I noticed the resizing of sliderspinners issue, for me the problem is only when they are made small

**Comment by Minecraftian14** [comment]
Yeah sure, while I don't want to go through every Issue listed here, if you think I am suitable for a specific task, I am ready!

My only issue are my exams, so there will be gaps where i cant even read the mails.😅

**Comment by lbalazscs** [comment]
It's not worth going through the open issues, because either they are already fixed (and they are still open only because the fix was not yet released in a non-beta version) or are pretty hard to fix. I plan to create special issues labeled "good first issue" for people like you.

> if you think I am suitable for a specific task, I am ready!

A problem is that I don't know much about your skills and interests, except that you love Swing, and you also are pretty good at it. I tried to come up with some ideas that don't require the understanding of a lot of existing code. Here they are, let me know if any of these sounds interesting:

1) Rectangular spiral. Currently there is a regular spiral (Filter/Render/Shapes/Spiral), but it would be interesting to also have a rectangular one (do a google image search for "Rectangular spiral" to see what I mean). Should be relatively easy and interesting if you like algorithms. The Spiral class should be modified.

2) Triangle/hex grid. Similarly, it would be nice if Filter/Render/Shapes/Grid could render hexagonal and triangle grids, not just the squares. This is in the RenderGrid class.

3) Drawing the viewed area in the Navigator. Currently the Navigator dialog allows the user to drag a red rectangle (called "view box") in order to move to a specific part of a zoomed-in image. It would be nice if the user could also ctrl-drag in the Navigator window in order to create an arbitrary view box, also changing the zoom level. This feature exists in Photoshop, a website describes it like this: "When using Photoshop's Navigator panel, the fastest way to zoom in on your image is by dragging a selection around the area you need. Press and hold your Ctrl (Win) / Command (Mac) key and drag out a selection (a View Box) around the area where you want to zoom in. Release your mouse button, and Photoshop will instantly zoom in to that area."
https://www.photoshopessentials.com/basics/photoshop-image-navigation-tips-tricks-and-shortcuts/

> My only issue are my exams, so there will be gaps where i cant even read the mails.

Don't worry about it. The worst thing that could happen is that somebody else implements these before you (I don't plan to work on these in the foreseeable future, but I might create the "good first issue" issues), but even if this happens, there will be other things to do...

**Comment by Minecraftian14** [comment]
I am doing 1, so please reserve it😆😅

**Pull request merged**